### ISSUE TITLE
URL roundtripping / redirect problems.

### ISSUE THREAD

**Issue by Ferada** [softwaremill/sttp#171]
See below for three test cases I'd really like to all pass with `sttp`.

Test three passes, but that's only because I explicitly set the encoding that way, test one and two are the ones I have problems with. Basically usage one is having a URL coming back from a backend as a string, so, to download it, it needs to be a `Uri`. Now, `java.net.URI` does this roundtrip, while mixing in `Uri` doesn't and that's trouble for anything that's not constructed via the API.

To make it worse, I also then had the problem that a redirect that occurred when I tried this failed completely because somehow during parsing the `Location` header the whole URL ended up in the `scheme` part of `Uri`, failing during validation. I suspect that's just a consequence of the original issue though, but again, as a library user I can't force the encoding override there, can I?

Am I missing something here with how to force the encoding in such a way that the original source is parsed and then represented exactly as passed in? I did not check any RFC for this, I'm looking at the behaviour of other libraries (and `curl`) only.

```
package example

import java.net.URI

import com.softwaremill.sttp.Uri
import com.softwaremill.sttp.Uri.{QueryFragment, QueryFragmentEncoding}
import org.scalatest._

class SttpUrlTests extends FlatSpec with Matchers {
  val source =
    "http://localhost:1234/a/b/c?token=token&url=http%3A%2F%2Fexample.com%3A9000%2Fsome-path%3Fparameter%3Dvalue"

  "sttp" should "roundtrip interesting URLs" in {
    val java = new URI(source)
    java.toString shouldBe source

    pendingUntilFixed {
      val scala = Uri(java)
      scala.toString shouldBe source
    }
  }

  "sttp" should "roundtrip its own interesting URLs with params" in {
    pendingUntilFixed {
      val scala = Uri("http", "localhost", 1234)
        .path("/a/b/c")
        .param("token", "token")
        .param("url", "http://example.com:9000/some-path?parameter=value")
      scala.toString shouldBe source
    }
  }

  "sttp" should "roundtrip its own interesting URLs with query fragments" in {
    val scala = Uri("http", "localhost", 1234)
      .path("/a/b/c")
      .queryFragment(QueryFragment.KeyValue("token", "token"))
      .queryFragment(
        QueryFragment.KeyValue("url",
                               "http://example.com:9000/some-path?parameter=value",
                               QueryFragmentEncoding.All,
                               QueryFragmentEncoding.All))
    scala.toString shouldBe source
  }
}
```

Edit: For case two, adding something like `defaultEncoding` to `Uri` and then setting it to `QueryFragmentEncoding.ALL` as a default value for all subsequent `param` calls would make this easier and backwards compatible with the current usage, of course just having to use `queryFragment` instead of `param` would also be an answer, though unconvenient.

For case one, I really don't know. Going from `URI.toString` through `UriInterpolator` means that the query will be parsed from scratch and the most obvious thing for me would be to set the encoding for all fragments to `QueryFragmentEncoding.ALL` created in `UriInterpolator` to keep it stable. The only test cases that break with that are the ones using `v3encoded` ... which I'm not sure what the right output there should be in the first place.

**Comment by Ferada** [comment]
Heya, thanks for the response, I suppose this might also be the server misbehaving. That said, for the redirect I made this small example:

```
package io.shiftleft.kraken

import com.softwaremill.sttp._

import scala.language.{higherKinds, postfixOps}
import scala.util.Try

object Derp extends App {
  val token = "..."
  val url = uri"http://localhost:8687/foo?token=...&url=http%3A%2F%2F..."

  implicit val backend: SttpBackend[Try, Nothing] = TryHttpURLConnectionBackend()

  val response = sttp.auth.bearer(token).get(url).send().get

  println(response)
}
```

Which then fails like this:

```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: Scheme can only contain alphanumeric characters, +, - and .
	at scala.Predef$.require(Predef.scala:277)
	at com.softwaremill.sttp.Uri.<init>(Uri.scala:37)
	at com.softwaremill.sttp.Uri.copy(Uri.scala:30)
	at com.softwaremill.sttp.Uri.scheme(Uri.scala:39)
	at com.softwaremill.sttp.UriInterpolator$UriBuilder$Scheme$.fromTokens(UriInterpolator.scala:266)
	at com.softwaremill.sttp.UriInterpolator$.$anonfun$interpolate$1(UriInterpolator.scala:24)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:122)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:118)
	at scala.collection.immutable.List.foldLeft(List.scala:85)
	at com.softwaremill.sttp.UriInterpolator$.interpolate(UriInterpolator.scala:22)
	at com.softwaremill.sttp.SttpApi$UriContext.uri(SttpApi.scala:169)
	at com.softwaremill.sttp.FollowRedirectsBackend.followRedirect(FollowRedirectsBackend.scala:53)
	at com.softwaremill.sttp.FollowRedirectsBackend.$anonfun$followRedirect$2(FollowRedirectsBackend.scala:37)
	at scala.Option.fold(Option.scala:158)
	at com.softwaremill.sttp.FollowRedirectsBackend.followRedirect(FollowRedirectsBackend.scala:33)
	at com.softwaremill.sttp.FollowRedirectsBackend.$anonfun$sendWithCounter$1(FollowRedirectsBackend.scala:21)
	at com.softwaremill.sttp.IdMonad$.flatMap(MonadError.scala:44)
	at com.softwaremill.sttp.FollowRedirectsBackend.sendWithCounter(FollowRedirectsBackend.scala:19)
	at com.softwaremill.sttp.FollowRedirectsBackend.send(FollowRedirectsBackend.scala:12)
	at com.softwaremill.sttp.TryBackend.$anonfun$send$1(TryBackend.scala:13)
	at scala.util.Try$.apply(Try.scala:209)
	at com.softwaremill.sttp.TryBackend.send(TryBackend.scala:13)
	at com.softwaremill.sttp.TryBackend.send(TryBackend.scala:11)
	...
Disconnected from the target VM, address: '127.0.0.1:44523', transport: 'socket'
```

Looking at the headers, the `Location` header is similar to this:

```
/foo%3Ftoken=...&url=http:/minio:9000/...?token=...&url=http://minio:9000/...?...
```

So, that looks _weird_. And it fails later because `isRelative` in `FollowRedirectsBackend.scala` considers the value as _not_ relative, because it contains literally `://` (not encoded).

---

I'm guessing this is a failure on the server part since the header looks really broken to me. Closing under that assumption.

**Comment by Ferada** [comment]
@adamw right I was wondering about it, but instead of the schema might be easier to check if `/` is the first character?

**Comment by adamw** [comment]
@Ferada should be fixed in 1.5.7

**Comment by Ferada** [comment]
@adamw thanks a lot for your help!
